### PR TITLE
Provide structured suggestion for `#![feature(foo)]`

### LINKED ISSUE #55941: Add X to crate attributes to enable message is not very clear
```
error[E0658]: use of unstable library feature 'try_from' (see issue #33417)
   --> src/schema.rs:441:9
    |
441 |     use std::convert::TryFrom;
    |         ^^^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(try_from)] to the crate attributes to enable
````

The message above is what I get attempting to use try_from. The issue is that it's not clear *where* crate attributes are, or whey they go. The rust book makes little mention of them. Adding this to "lib.rs",  causes new errors such as:

```
error: an inner attribute is not permitted in this context
  --> src/lib.rs:15:3
   |
15 | #![feature(try_from)]
   |   ^
   |
   = note: inner attributes, like `#![no_std]`, annotate the item enclosing them, and are usually found at the beginning of source files. Outer attributes, like `#[test]`, annotate the item following them.

````

As a result, it's really not clear how to resolve this error.

The descriptive error rustc --explain E0658 also doesn't demonstrate what to do correctly. 

The error message should explain *which* file to add the #[feature] line too. Additionally, the E0658 should explain not just what file, but what to do in a multi-file case (ie lib.rs contains mod foo, and foo.rs wants to use the feature).

This would help to make this easier to resolve and access these features. 

EDIT: It would be also useful to handle the mod tests situation as well where you want to test a feature like TryFrom in mod tests, because that likely needs different/other handling?

### PR DESCRIPTION
**Pull request by estebank**
```
error: `S2<'_>` is forbidden as the type of a const generic parameter
  --> $DIR/lifetime-in-const-param.rs:5:23
   |
LL | struct S<'a, const N: S2>(&'a ());
   |                       ^^
   |
   = note: the only supported types are integers, `bool` and `char`
help: add `#![feature(adt_const_params)]` to the crate attributes to enable more complex and user defined types
   |
LL + #![feature(adt_const_params)]
   |
```

Fix #55941.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -2104,6 +2106,44 @@ impl<'tcx> TyCtxt<'tcx> {
...
+            }
+        }
+        None
+    }
+
+    pub fn disabled_nightly_feature<E: rustc_errors::EmissionGuarantee>(
+        self,
+        diag: &mut Diag<'_, E>,
+        hir_id: Option<HirId>,
+        features: impl Iterator<Item = (String, Symbol)>,
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/ty/context.rs:2124]:
```suggestion
        features: impl IntoIterator<Item = (String, Symbol)>,
```

Given that you use `.into_iter()` in basically all the calls

File: compiler/rustc_const_eval/src/transform/check_consts/ops.rs [hunk 2]
```diff
@@ -297,10 +297,12 @@ impl<'tcx> NonConstOp<'tcx> for FnCallNonConst<'tcx> {
...
 
-        if let Some(feature) = feature
-            && ccx.tcx.sess.is_nightly_build()
-        {
-            err.help(format!("add `#![feature({feature})]` to the crate attributes to enable",));
+        if let Some(feature) = feature {
+            ccx.tcx.disabled_nightly_feature(
+                &mut err,
+                body.source.def_id().as_local().map(|local| ccx.tcx.local_def_id_to_hir_id(local)),
+                vec![(String::new(), feature)].into_iter(),
```
Comment (WaffleLapkin) [inline compiler/rustc_const_eval/src/transform/check_consts/ops.rs:304]:
```suggestion
                [(String::new(), feature)],
```
tip: you don't need vectors to create iterators, arrays implement `IntoIterator` :)

File: compiler/rustc_middle/src/ty/context.rs [hunk 3]
```diff
@@ -2104,6 +2106,44 @@ impl<'tcx> TyCtxt<'tcx> {
         lint_level(self.sess, lint, level, src, Some(span.into()), msg, decorate);
     }
 
+    /// Find the crate root and the appropriate span where `use` and outer attributes can be
+    /// inserted at.
+    pub fn crate_inject_span(self, hir_id: HirId) -> Option<Span> {
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/ty/context.rs:2111]: The name is not very clear to me... maybe something like `crate_level_attribute_injection_span`?

Cause the current version I read as "inject a span into a crate"...

File: compiler/rustc_middle/src/ty/context.rs [hunk 4]
```diff
@@ -2104,6 +2106,44 @@ impl<'tcx> TyCtxt<'tcx> {
...
+        features: impl Iterator<Item = (String, Symbol)>,
+    ) {
+        if !self.sess.is_nightly_build() {
+            return;
+        }
+
+        let span = hir_id.and_then(|id| self.crate_inject_span(id));
+        for (desc, feature) in features {
+            let msg =
+                format!("add `#![feature({feature})]` to the crate attributes to enable{desc}");
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/ty/context.rs:2133]: Can this be made translatable too? (or should we at least leave a fixme?)

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
r? WaffleLapkin

File: compiler/rustc_middle/src/ty/context.rs (hunk 4 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/context.rs:2201]: left a `FIXME` :)

File: compiler/rustc_middle/src/ty/context.rs [hunk 5]
```diff
@@ -2171,6 +2173,45 @@ impl<'tcx> TyCtxt<'tcx> {
...
+    pub fn crate_level_attribute_injection_span(self, hir_id: HirId) -> Option<Span> {
+        for (_hir_id, node) in self.hir().parent_iter(hir_id) {
+            if let hir::Node::Crate(m) = node {
+                return Some(m.spans.inject_use_span.shrink_to_lo());
+            }
+        }
+        None
+    }
+
+    pub fn disabled_nightly_feature<E: rustc_errors::EmissionGuarantee>(
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/ty/context.rs:2187]:
```suggestion
    pub fn disabled_nightly_features<E: rustc_errors::EmissionGuarantee>(
```

**Review by WaffleLapkin — APPROVED** [review]
r=me with the nit

**Pull request merged**